pkg/grpc_middleware/dialer/builder.go

```diff
@@ -0,0 +1,319 @@
+package dialer
+
+import (
+	"context"
+	"crypto/tls"
+	"crypto/x509"
+	"fmt"
+	"log"
+	"net/url"
+	"strconv"
+	"time"
+
+	"github.com/caring/go-packages/v2/pkg/errors"
+	grpc_middleware "github.com/grpc-ecosystem/go-grpc-middleware"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/backoff"
+	"google.golang.org/grpc/credentials"
+	"google.golang.org/grpc/keepalive"
+)
+
+const (
+	// MinConnectTimeout before attempting reconnect.
+	MinConnectTimeout = 20 * time.Second
+)
+
```

Comment:
I like getting rid of the interface, but IIRC if you pass in an interface into a method, you can only call methods defined on the interface within that method, and not methods defined on a type that implements that interface... in that case, we should just pass around the concrete type, and probably need to make a standalone repo for these packages to make cross-repo compatibility easier...

Reply:
Does that mean in every use case, the engineer needs to create a custom interface to define the methods they plan on using from the builder in their situation?

Reply:
They don't *HAVE* to.. i had to in this instance because of the conflicting packages between my custom version of go-packages and the current go-packages master.